### PR TITLE
use Last() implementation when calculating tournament deadlines

### PR DESCRIPTION
**Pull request by lugehorsam**
Ports https://github.com/gorhill/cronexpr/pull/28 to our internal version of `cronexpr`. This gives us the `Last()` method which allows us to infer the start time directly from the cronexpr rather than estimating it by invoking `Next()` twice.

- Fixes calculations of leaderboard and tournament times for rare cron expressions that don't execute at a fixed interval.
- Fixes calculations of leaderboard and tournament times for future tournaments.

### CODE REVIEW

File: server/core_tournament.go [hunk 1]
```diff
@@ -737,22 +737,17 @@ func calculateTournamentDeadlines(startTime, endTime, duration int64, resetSched
...
-					startActiveUnix = s
-				}
-				break
-			}
+		var startActiveUnix int64
+		// check if we are landing squarely on the reset schedule
+		if resetSchedule.Next(t.Add(-1*time.Second)) == t {
+			startActiveUnix = tUnix
+		} else {
+			// otherwise assume the supplied time is ahead of the reset schedule.
```
Comment (lugehorsam) [inline server/core_tournament.go:745]: I believe this assumption was baked into the for loop of the previous code as well.

**Comment by lugehorsam** [comment]
I added a test for an additional case I thought of but this is still good to review.

File: server/core_tournament.go (hunk 1 above)
Comment (sesposito) [reply server/core_tournament.go:745]: Is this covering the case where `startActiveTime` is in the future?

File: server/core_tournament_test.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+	startActiveUnix, endActiveUnix, _ := calculateTournamentDeadlines(startTime, 0, duration, sched, nowUnix)
+	nextResetUnix := sched.Next(nowUnix).Unix()
+
+	// 14 August 2023, 9:00:00
+	require.Equal(t, int64(1692003600), startActiveUnix, "Start active times should be equal.")
+	// 21 August 2023, 9:00:00
+	require.Equal(t, int64(1692608400), endActiveUnix, "End active times should be equal.")
+	// 14 September 2023, 9:00:00
+	require.Equal(t, int64(1694682000), nextResetUnix, "Next reset times should be equal.")
+}
```
Comment (sesposito) [inline server/core_tournament_test.go:92]: I'd suggest adding a test case for when startTime > now

File: server/core_tournament.go (hunk 1 above)
Comment (lugehorsam) [reply server/core_tournament.go:745]: Seeing the leaderboard code, it looks like `0` should be returned if `startActiveTime` is in the future.

**Comment by lugehorsam** [comment]
Ready for re-review.

**Review by syhpoon — APPROVED** [review]
:+1:

**Comment by lugehorsam** [comment]
Will still need a 👍 from @sesposito  if it deserves it to unblock the merge

**Review by sesposito — APPROVED** [review]

**Pull request merged**